server/src/test/java/org/elasticsearch/common/io/stream/BytesStreamOutputTests.java

```diff
@@ -0,0 +1,123 @@
+
+    private BytesStreamOutput stream;
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        stream = new BytesStreamOutput();
+    }
+
+    public void testDefaultConstructor() {
+        assertEquals(0, stream.size());
+        assertEquals(0, stream.position());
+        assertEquals(BytesArray.EMPTY, stream.bytes());
+    }
+
+    public void testConstructorWithExpectedSize() {
+        BytesStreamOutput s = new BytesStreamOutput(randomNonNegativeInt());
+        assertEquals(0, s.size());
+        assertEquals(0, s.position());
+    }
+
+    public void testWriteByte() {
+        byte i1 = randomByte();
+        byte i2 = randomByte();
+        stream.writeByte(i1);
+        stream.writeByte(i2);
+        assertEquals(2, stream.size());
+        assertEquals(2, stream.position());
+        assertEquals(i1, stream.bytes().get(0));
+        assertEquals(i2, stream.bytes().get(1));
+    }
+
+    public void testWriteBytes() {
+        int maxSize = randomIntBetween(10, 100);
+        byte[] data = randomByteArrayOfLength(maxSize);
+        int length = randomIntBetween(0, maxSize);
+        stream.writeBytes(data, 0, length);
+        assertEquals(length, stream.size());
+        for (int i = 0; i < length; i++) {
+            assertEquals(data[i], stream.bytes().get(i));
+        }
+    }
+
+    public void testWriteBytesWithOffset() {
+        int maxSize = randomIntBetween(10, 100);
+        byte[] data = randomByteArrayOfLength(maxSize);
+        int offset = randomIntBetween(0, maxSize);
+        int length = randomIntBetween(0, maxSize - offset);
+        stream.writeBytes(data, offset, length);
+        assertEquals(length, stream.size());
+        for (int i = 0; i < stream.size(); i++) {
+            assertEquals(stream.bytes().get(i), data[i + offset]);
+        }
+    }
+
+    public void testWriteBytesWithLengthGreaterThanData() {
+        BytesStreamOutput s = new BytesStreamOutput(0, BigArrays.NON_RECYCLING_INSTANCE);
+        int size = randomIntBetween(0, 100);
+        assertThrows(IndexOutOfBoundsException.class, () -> s.writeBytes(randomByteArrayOfLength(size), 0, size + 1));
+    }
+
+    public void testResetShrinksAndResetsCount() {
+        int byteArrayLength = randomIntBetween(0, 100);
+        stream.writeBytes(randomByteArrayOfLength(byteArrayLength), 0, byteArrayLength);
+        stream.reset();
+        assertEquals(0, stream.size());
+        // After reset, should be able to write again
+        byte b = randomByte();
+        stream.writeByte(b);
+        assertEquals(1, stream.size());
+        assertEquals(b, stream.bytes().get(0));
+    }
+
+    public void testSeekAndSkip() {
+        int byteArrayLength = randomIntBetween(0, 100);
+        byte[] byteArray = randomByteArrayOfLength(byteArrayLength);
+        stream.writeBytes(byteArray, 0, byteArrayLength);
+
+        int seekPosition = randomIntBetween(byteArrayLength, byteArrayLength + 100);
```

Comment:
IIUC, a seek could be done to somewhere between 0 and `byteArrayLength`, and then a read can be done to show that it matches the `byteArray`? Same thing with skip below, though the seek would have to be further constrained to leave some room to test skip on the data set.

Would that be more realistic than setting the read position into space that hasn't been written yet? Or maybe test both?